nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,1115 @@
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims.insert(index, value)
+
+    def _to_xml_element(self):
+        if (len(self) == 0 and self.metadata is None):
+            raise CIFTI2HeaderError(
+                'Matrix element requires either a MatrixIndicesMap or a Metadata element'
+            )
+
+        mat = xml.Element('Matrix')
+        if self.metadata:
+            mat.append(self.metadata._to_xml_element())
+        for mim in self._mims:
+            mat.append(mim._to_xml_element())
+        return mat
+
+
+class Cifti2Header(FileBasedHeader, xml.XmlSerializable):
+    ''' Class for Cifti2 header extension '''
+
+    def __init__(self, matrix=None, version="2.0"):
+        FileBasedHeader.__init__(self)
+        xml.XmlSerializable.__init__(self)
+        self.matrix = Cifti2Matrix() if matrix is None else Cifti2Matrix()
+        self.version = version
+
+    def _to_xml_element(self):
+        cifti = xml.Element('CIFTI')
+        cifti.attrib['Version'] = str(self.version)
+        mat_xml = self.matrix._to_xml_element()
+        if mat_xml is not None:
+            cifti.append(mat_xml)
+        return cifti
+
+    @classmethod
+    def from_header(klass, header=None):
+        if header is None:
+            return klass()
+        if type(header) == klass:
+            return header.copy()
+        raise ValueError('header is not a Cifti2Header')
+
+    @classmethod
+    def may_contain_header(klass, binaryblock):
+        from .parse_cifti2_fast import _Cifti2AsNiftiHeader
+        return _Cifti2AsNiftiHeader.may_contain_header(binaryblock)
+
+
+class Cifti2Image(FileBasedImage):
+    """ Class for single file CIfTI2 format image
+    """
+    header_class = Cifti2Header
+    valid_exts = Nifti2Image.valid_exts
+    files_types = Nifti2Image.files_types
+    makeable = False
+    rw = True
+
+    def __init__(self, data=None, header=None, nifti_header=None):
+        ''' Initialize image
+
+        The image is a combination of (array, affine matrix, header, nifti_header),
+        with optional metadata in `extra`, and filename / file-like objects
+        contained in the `file_map` mapping.
+
+        Parameters
+        ----------
+        dataobj : object
+           Object containg image data.  It should be some object that retuns an
+           array from ``np.asanyarray``.  It should have a ``shape`` attribute
+           or property
+        affine : None or (4,4) array-like
+           homogenous affine giving relationship between voxel coordinates and
+           world coordinates.  Affine can also be None.  In this case,
+           ``obj.affine`` also returns None, and the affine as written to disk
+           will depend on the file format.
+        header : Cifti2Header object
+        nifti_header : None or mapping or nifti2 header instance, optional
+           metadata for this image format
+        '''
+        self._header = header or Cifti2Header()
```

Comment:
Need to check if `header` is valid?  Maybe `self._header = Cifti2Header() if header is None else Cifti2Header.from_header(header)`.